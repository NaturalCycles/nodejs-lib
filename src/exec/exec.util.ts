import c from 'chalk'
import * as execa from 'execa'

export interface ExecaOptions extends execa.Options {
  /**
   * If true - it will reject a promise with an error and NOT do `process.exit`
   */
  noProcessExit?: boolean
}

export async function proxyCommand(
  cmd: string,
  args: string[] = [],
  opt: ExecaOptions = {},
): Promise<void> {
  const [, , ...processArgs] = process.argv

  await execCommand(cmd, [...args, ...processArgs], {
    ...opt,
  })
}

export async function execCommand(
  cmd: string,
  args: string[] = [],
  opt: ExecaOptions = {},
): Promise<void> {
  logExec(cmd, args, opt)

  await execa(cmd, args, {
    stdio: 'inherit',
    preferLocal: true,
    ...opt,
  }).catch(err => {
    if (opt.noProcessExit) {
      throw err || new Error(`execCommand failed: ${cmd}`)
    }

    if (err) {
      if (err.exitCodeName === 'ENOENT') {
        console.log(`Error: ENOENT (no such file or directory): ${cmd}`)
        process.exit(1)
      } else if (err.exitCode) {
        process.exit(err.exitCode)
      } else if (err.exitCodeName) {
        console.log(`Error: ${err.exitCodeName}`)
      }
    }

    process.exit(1)
  })
}

/**
 * Convenience method that calls execCommand with `shell: true` option.
 */
export async function execShell(cmd: string, opt: ExecaOptions = {}): Promise<void> {
  await execCommand(cmd, [], {
    shell: true,
    ...opt,
  })
}

export function logExec(cmd: string, args: string[] = [], opt: ExecaOptions = {}): void {
  const cmdline = [
    ...Object.entries(opt.env || {}).map(([k, v]) => [k, v].join('=')),
    cmd,
    ...args,
  ].join(' ')

  console.log(c.grey(cmdline))
}
