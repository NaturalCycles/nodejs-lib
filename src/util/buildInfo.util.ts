import {
  _filterUndefinedValues,
  AnyObject,
  BuildInfo,
  localTimeOrNow,
  UnixTimestampNumber,
} from '@naturalcycles/js-lib'
import { fs2 } from '../fs/fs2'
import {
  gitCurrentBranchName,
  gitCurrentCommitSha,
  gitCurrentCommitTimestamp,
  gitCurrentRepoName,
} from './git.util'

export interface GenerateBuildInfoOptions {
  /**
   * If set - this timestamp will be used, instead of "current time".
   */
  overrideTimestamp?: UnixTimestampNumber
}

export function generateBuildInfo(opt: GenerateBuildInfoOptions = {}): BuildInfo {
  const now = localTimeOrNow(opt.overrideTimestamp)
  const ts = now.unix()

  const rev = gitCurrentCommitSha()
  const branchName = gitCurrentBranchName()
  const repoName = gitCurrentRepoName()
  const tsCommit = gitCurrentCommitTimestamp()

  const ver = [now.toStringCompact(), repoName, branchName, rev].join('_')

  let { APP_ENV: env } = process.env

  if (!env) {
    // Attempt to read `envByBranch` from package.json root
    try {
      if (fs2.pathExists('package.json')) {
        const packageJson = fs2.readJson<AnyObject>('package.json')
        env = packageJson?.['envByBranch']?.[branchName] || packageJson?.['envByBranch']?.['*']
      }
    } catch {}
  }

  return _filterUndefinedValues({
    ts,
    tsCommit,
    repoName,
    branchName,
    rev,
    ver,
    env,
  })
}
