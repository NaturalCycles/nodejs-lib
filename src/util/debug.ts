import debug from 'debug'
import { IDebugger } from 'debug'
import * as path from 'path'

export function getDebug (filename: string): IDebugger {
  return debug(getDebugNamespaceFromFilename(filename))
}

function getDebugNamespaceFromFilename (filename: string): string {
  // drop last part after dot
  const a = path.basename(filename).split('.')
  a.pop()
  return a.join('.')
}
