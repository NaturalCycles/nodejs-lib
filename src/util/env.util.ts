import type { ValuesOf } from '@naturalcycles/js-lib'
import 'dotenv/config' // ensure .env is read before requiring keys
import * as fs from 'fs-extra'

/**
 * @example
 *
 * const {a, b} = requreEnvKeys(['a', 'b'])
 *
 * Will throw if any of the passed keys is not defined.
 */
export function requireEnvKeys<T extends readonly string[]>(
  ...keys: T
): { [k in ValuesOf<T>]: string } {
  // eslint-disable-next-line unicorn/no-array-reduce
  return keys.reduce((r, k) => {
    const v = process.env[k]
    if (!v) throw new Error(`${k} env variable is required, but missing`)
    r[k] = v
    return r
  }, {} as { [k in ValuesOf<T>]: string })
}

export function requireFileToExist(filePath: string): void {
  if (!fs.pathExistsSync(filePath)) {
    throw new Error(`Required file should exist: ${filePath}`)
  }
}
