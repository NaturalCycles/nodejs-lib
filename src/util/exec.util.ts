import type { ProcessEnvOptions, SpawnOptions } from 'node:child_process'
import * as cp from 'node:child_process'

export async function execVoidCommand(
  cmd: string,
  args: string[] = [],
  opt: SpawnOptions = {},
): Promise<void> {
  logExec(cmd, args, opt)

  await new Promise<void>(resolve => {
    const p = cp.spawn(cmd, [...args], {
      stdio: 'inherit',
      // shell: true,
      ...opt,
      env: {
        ...process.env,
        ...opt.env,
      },
    })

    p.on('close', code => {
      if (!code) return resolve()
      console.log(`${cmd} exited with code ${code}`)
      process.exit(code)
    })
  })
}

export function execVoidCommandSync(
  cmd: string,
  args: string[] = [],
  opt: SpawnOptions = {},
): void {
  logExec(cmd, args, opt)

  const r = cp.spawnSync(cmd, [...args], {
    encoding: 'utf8',
    stdio: 'inherit',
    // shell: true, // removing shell breaks executing `tsc`
    ...opt,
    env: {
      ...process.env,
      ...opt.env,
    },
  })

  if (r.status) {
    console.log(`${cmd} exited with code ${r.status}`)
    process.exit(r.status)
  }

  if (r.error) {
    console.log(r.error)
    process.exit((r.error as NodeJS.ErrnoException).errno || 1)
  }
}

function logExec(cmd: string, args: string[] = [], opt: ProcessEnvOptions = {}): void {
  const cmdline = [
    ...Object.entries(opt.env || {}).map(([k, v]) => [k, v].join('=')),
    cmd,
    ...args,
  ].join(' ')

  console.log(cmdline)
}
