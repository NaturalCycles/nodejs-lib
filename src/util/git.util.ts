import * as cp from 'node:child_process'
import * as path from 'node:path'
import type { UnixTimestampNumber } from '@naturalcycles/js-lib'
import { grey } from '../colors/colors'

export function getLastGitCommitMsg(): string {
  return execSync('git log -1 --pretty=%B')
}

export function commitMessageToTitleMessage(msg: string): string {
  const firstLine = msg.split('\n')[0]!
  const [preTitle, title] = firstLine.split(': ')
  return title || preTitle!
}

export function gitHasUncommittedChanges(): boolean {
  // git diff-index --quiet HEAD -- || echo "untracked"
  try {
    cp.execSync('git diff-index --quiet HEAD --', {
      encoding: 'utf8',
    })
    return false
  } catch {
    return true
  }
}

/**
 * @returns true if there were changes
 */
export function gitCommitAll(msg: string): boolean {
  // git commit -a -m "style(lint-all): $GIT_MSG" || true
  const cmd = `git commit -a --no-verify -m "${msg}"`
  // const cmd = `git`
  // const args = ['commit', '-a', '--no-verify', '-m', msg]

  console.log(grey(cmd))

  try {
    cp.execSync(cmd, {
      stdio: 'inherit',
    })
    return true
  } catch {
    return false
  }
}

/**
 * @returns true if there are not pushed commits.
 */
export function gitIsAhead(): boolean {
  // ahead=`git rev-list HEAD --not --remotes | wc -l | awk '{print $1}'`
  const cmd = `git rev-list HEAD --not --remotes | wc -l | awk '{print $1}'`

  const stdout = execSync(cmd)

  // console.log(`gitIsAhead: ${stdout}`)
  return Number(stdout) > 0
}

export function gitPull(): void {
  const cmd = 'git pull'
  try {
    cp.execSync(cmd, {
      stdio: 'inherit',
    })
  } catch {}
}

export function gitPush(): void {
  // git push --set-upstream origin $CIRCLE_BRANCH && echo "pushed, exiting" && exit 0
  let cmd = 'git push'

  const branchName = gitCurrentBranchName()

  if (branchName) {
    cmd += ` --set-upstream origin ${branchName}`
  }

  console.log(grey(cmd))

  cp.execSync(cmd, {
    stdio: 'inherit',
  })
}

export function gitCurrentCommitSha(full = false): string {
  const sha = execSync('git rev-parse HEAD')
  return full ? sha : sha.slice(0, 7)
}

export function gitCurrentCommitTimestamp(): UnixTimestampNumber {
  return Number(execSync('git log -1 --format=%ct'))
}

export function gitCurrentBranchName(): string {
  return execSync('git rev-parse --abbrev-ref HEAD')
}

export function gitCurrentRepoName(): string {
  const originUrl = execSync('git config --get remote.origin.url')
  return path.basename(originUrl, '.git')
}

function execSync(cmd: string): string {
  return cp
    .execSync(cmd, {
      encoding: 'utf8',
      // stdio: 'inherit', // no, otherwise we don't get the output returned
    })
    .trim()
}
