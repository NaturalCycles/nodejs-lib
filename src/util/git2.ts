import cp from 'node:child_process'
import path from 'node:path'
import type { UnixTimestampNumber } from '@naturalcycles/js-lib'
import { grey } from '../colors/colors'
import { exec2 } from './exec2'

/**
 * Set of utility functions to work with git.
 */
class Git2 {
  getLastGitCommitMsg(): string {
    return exec2.exec('git log -1 --pretty=%B')
  }

  commitMessageToTitleMessage(msg: string): string {
    const firstLine = msg.split('\n')[0]!
    const [preTitle, title] = firstLine.split(': ')
    return title || preTitle!
  }

  gitHasUncommittedChanges(): boolean {
    // git diff-index --quiet HEAD -- || echo "untracked"
    try {
      cp.execSync('git diff-index --quiet HEAD --', {
        encoding: 'utf8',
      })
      return false
    } catch {
      return true
    }
  }

  /**
   * Returns true if there were changes
   */
  gitCommitAll(msg: string): boolean {
    // git commit -a -m "style(lint-all): $GIT_MSG" || true
    const cmd = `git commit -a --no-verify -m "${msg}"`
    // const cmd = `git`
    // const args = ['commit', '-a', '--no-verify', '-m', msg]
    console.log(grey(cmd))

    try {
      cp.execSync(cmd, {
        stdio: 'inherit',
      })
      return true
    } catch {
      return false
    }
  }

  /**
   * @returns true if there are not pushed commits.
   */
  gitIsAhead(): boolean {
    // ahead=`git rev-list HEAD --not --remotes | wc -l | awk '{print $1}'`
    const cmd = `git rev-list HEAD --not --remotes | wc -l | awk '{print $1}'`
    const stdout = exec2.exec(cmd)
    // console.log(`gitIsAhead: ${stdout}`)
    return Number(stdout) > 0
  }

  gitPull(): void {
    const cmd = 'git pull'
    try {
      cp.execSync(cmd, {
        stdio: 'inherit',
      })
    } catch {}
  }

  gitPush(): void {
    // git push --set-upstream origin $CIRCLE_BRANCH && echo "pushed, exiting" && exit 0
    let cmd = 'git push'

    const branchName = this.gitCurrentBranchName()

    if (branchName) {
      cmd += ` --set-upstream origin ${branchName}`
    }

    exec2.spawn(cmd, { logStart: true })
  }

  gitCurrentCommitSha(full = false): string {
    const sha = exec2.exec('git rev-parse HEAD')
    return full ? sha : sha.slice(0, 7)
  }

  gitCurrentCommitTimestamp(): UnixTimestampNumber {
    return Number(exec2.exec('git log -1 --format=%ct'))
  }

  gitCurrentBranchName(): string {
    return exec2.exec('git rev-parse --abbrev-ref HEAD')
  }

  gitCurrentRepoName(): string {
    const originUrl = exec2.exec('git config --get remote.origin.url')
    return path.basename(originUrl, '.git')
  }
}

export const git2 = new Git2()
