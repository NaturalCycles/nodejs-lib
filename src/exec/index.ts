import { ExecaOptions, execCommand, execShell, execWithArgs } from './exec.util'

export type { ExecaOptions }

export { execCommand, execShell, execWithArgs }
