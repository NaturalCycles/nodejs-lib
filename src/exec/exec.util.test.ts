import { execCommand } from './exec.util'

beforeEach(() => {
  jest.restoreAllMocks()
})

test('execCommand ok', async () => {
  await execCommand('ls > /dev/null', [], {
    shell: true,
  })
})

test('execCommand error, exit', async () => {
  const processExit = jest.spyOn(process, 'exit').mockImplementation(code => {
    throw new Error(`process exit: ${code}`)
  })
  await expect(execCommand('abcd')).rejects.toThrow()
  expect(processExit).toMatchSnapshot()
})
