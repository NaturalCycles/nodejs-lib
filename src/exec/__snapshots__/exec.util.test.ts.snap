// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execCommand error, exit 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      1,
    ],
  ],
  "results": Array [
    Object {
      "type": "throw",
      "value": [Error: process exit: 1],
    },
  ],
}
`;
